payments-core/src/main/java/com/stripe/android/payments/core/authentication/SourceAuthenticator.kt

```diff
@@ -0,0 +1,79 @@
+package com.stripe.android.payments.core.authentication
+
+import com.stripe.android.PaymentBrowserAuthStarter
+import com.stripe.android.PaymentRelayStarter
+import com.stripe.android.StripePaymentController
+import com.stripe.android.auth.PaymentBrowserAuthContract
+import com.stripe.android.model.Source
+import com.stripe.android.networking.AnalyticsEvent
+import com.stripe.android.networking.AnalyticsRequestExecutor
+import com.stripe.android.networking.AnalyticsRequestFactory
+import com.stripe.android.networking.ApiRequest
+import com.stripe.android.payments.core.injection.ENABLE_LOGGING
+import com.stripe.android.payments.core.injection.UIContext
+import com.stripe.android.view.AuthActivityStarterHost
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+import kotlin.coroutines.CoroutineContext
+
+@Singleton
+internal class SourceAuthenticator @Inject constructor(
+    private val paymentBrowserAuthStarterFactory: (AuthActivityStarterHost) -> PaymentBrowserAuthStarter,
+    private val paymentRelayStarterFactory: (AuthActivityStarterHost) -> PaymentRelayStarter,
+    private val analyticsRequestExecutor: AnalyticsRequestExecutor,
+    private val analyticsRequestFactory: AnalyticsRequestFactory,
+    @Named(ENABLE_LOGGING) private val enableLogging: Boolean,
+    @UIContext private val uiContext: CoroutineContext
+) : PaymentAuthenticator {
+    override suspend fun authenticate(
+        host: AuthActivityStarterHost,
+        authenticatable: Authenticatable,
+        requestOptions: ApiRequest.Options
+    ) {
+        val source = authenticatable as Source
```

Comment:
Can we use generics on `PaymentAuthenticator` to specify the `Authenticatable` concrete type?

Reply:
Great suggestion 👍 Just deleted the `Authenticatable` interface and was able to clean up a lot of downcasting.
We do need some unchecked cast in `DefaultPaymentAuthenticatorRegistry#getAuthenticator`, but it's still way less verbose compared to a dedicate interface.